src/AppInstallerRepositoryCore/RepositorySource.cpp

```diff
@@ -578,36 +584,44 @@ namespace AppInstaller::Repository
     std::shared_ptr<ISource> OpenPredefinedSource(PredefinedSource source, IProgressCallback& progress)
     {
         SourceDetails details;
+        details.Origin = SourceOrigin::Predefined;
 
         switch (source)
         {
         case PredefinedSource::Installed:
-            // TODO: Pull directly from factory
-            details.Type = "Microsoft.Predefined.Installed";
+            details.Type = Microsoft::PredefinedInstalledSourceFactory::Type();
+            details.Arg = Microsoft::PredefinedInstalledSourceFactory::FilterToString(Microsoft::PredefinedInstalledSourceFactory::Filter::None);
             return CreateSourceFromDetails(details, progress);
         case PredefinedSource::ARP_System:
-            // TODO: Pull directly from factory
-            details.Type = "Microsoft.Predefined.ARP";
-            details.Arg = "system";
+            details.Type = Microsoft::PredefinedInstalledSourceFactory::Type();
+            details.Arg = Microsoft::PredefinedInstalledSourceFactory::FilterToString(Microsoft::PredefinedInstalledSourceFactory::Filter::ARP_System);
             return CreateSourceFromDetails(details, progress);
         case PredefinedSource::ARP_User:
-            // TODO: Pull directly from factory
-            details.Type = "Microsoft.Predefined.ARP";
-            details.Arg = "user";
+            details.Type = Microsoft::PredefinedInstalledSourceFactory::Type();
+            details.Arg = Microsoft::PredefinedInstalledSourceFactory::FilterToString(Microsoft::PredefinedInstalledSourceFactory::Filter::ARP_User);
             return CreateSourceFromDetails(details, progress);
         case PredefinedSource::MSIX:
-            // TODO: Pull directly from factory
-            details.Type = "Microsoft.Predefined.MSIX";
+            details.Type = Microsoft::PredefinedInstalledSourceFactory::Type();
+            details.Arg = Microsoft::PredefinedInstalledSourceFactory::FilterToString(Microsoft::PredefinedInstalledSourceFactory::Filter::MSIX);
             return CreateSourceFromDetails(details, progress);
         }
 
         THROW_HR(E_UNEXPECTED);
     }
 
-    std::shared_ptr<ISource> CreateCompositeSource(std::shared_ptr<ISource>& source1, std::shared_ptr<ISource>&)
+    std::shared_ptr<ISource> CreateCompositeSource(const std::shared_ptr<ISource>& installedSource, const std::shared_ptr<ISource>& availableSource)
     {
-        // TODO: needs implementation
-        return source1;
+        std::shared_ptr<CompositeSource> result = std::dynamic_pointer_cast<CompositeSource>(availableSource);
+
+        if (!result)

```

Comment:
>!result [](start = 12, length = 7)

In what case will this happen?

Reply:
If there are multiple sources configured and no specific one is selected on the command line, then the incoming source should be Composite already.  If there is only one source configured, or a single one is selected via `--source`, then it won't be Composite.

---
In reply to: [507556019](https://github.com/microsoft/winget-cli/pull/598#discussion_r507556019) [](ancestors = 507556019)